
syntax = "proto3";


message Player {
	
	uint64 id = 1;
	string username = 2;
	uint32 wins = 3;
	uint32 tieds = 4;
	uint32 lost = 5;

}



// Operations
// 
message Operation {

	enum Request {
		RANKING = 0; // show ranking
		PLAY = 1; // play a game
		ONLINE = 2; // online users
		VIEW = 3; // view yourself information
		CHANGE = 4; // request updates in your acount
	}

	Request operation = 1;

}


message ResponseStatus {
	int32 status_code = 1;
	string message = 2;
}


// Login and Signup has the same request body
// 
message LogInSignUp {
	string username = 1;
	string password = 2;

	enum Type {
		SIGNUP = 0;
		LOGIN = 1;
	}

	Type request_type = 3;
}

// Login and Signup has the same reponse
// 
message LogInSignUpResponse {
	ResponseStatus status = 1;
	Player player = 2;
}


// Ranking
// 
message Ranking {
	ResponseStatus status = 1;
	repeated Player players = 2;
}

// Play
// 
message Play {

	enum Type {
		WAIT = 0;
		CHALLANGE = 1;
	}

	Type request_type = 1;


	Player challanded_player = 2;
}

// Challange
message ChallangeRequest {
	Player challander_player = 1;
}

message ChallangeResponse {
	enum Answer {
		ACCEPT = 0;
		DENY = 1;
	}

	Answer answer = 1;
}





// Online Users
// 
message OnlinePlayers {
	ResponseStatus status = 1;
	repeated Player pleyers = 2;
}



// Acount Changes
// 
message Change {
	string username = 1;
	string password = 2;
}

message ChangesResponse {
	ResponseStatus status = 1;

}
